### PR TITLE
[Bugfix] - Save Complaint after Instructor overwrote assessment + Server & Client Tests

### PR DESCRIPTION
**Pull request by FuchsDominik**
Fixes https://github.com/ls1intum/Artemis/issues/2026
Fixes https://github.com/ls1intum/Artemis/issues/2129

### Checklist
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)

### Motivation and Context

- Weekly client side test coverage improvements
- When overwriting an assessment as instructor after a complaint was created for this submission, the complaint is not shown anymore

### Description

- Get the client test coverage higher 🚀 
- Integration test covering the problem mentioned in https://github.com/ls1intum/Artemis/issues/2026 and https://github.com/ls1intum/Artemis/issues/2129 for modeling exercises and for programming exercises.
- Fixed a Bug where the complaint is overwritten in programming exercises
- Added some more delay on one test which did not pass once (Should fail less often since the test/call has more time to be executed)

### Steps for Testing

1. Log in to Artemis
2. Create a Programming-Exercise
3. Participate
4. Assess
5. Complain
6. (Accept/Reject the complaint) - Step not necessarily needed
7. As a instructor, override the assessment
8. The complaint should still be visible in the assessment view (same view when clicking in _override Assessment_)

### Test Coverage

short-answer-question.component.ts: 0% -> 94%

### CODE REVIEW

**Review by stefanwaldhauser — DISMISSED** [review]
Nice work :) Looks like a solid test

File: src/test/javascript/spec/component/short-answer-question/short-answer-question-edit.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+    it('should add spot at cursor visual mode', () => {
+        const textParts = [['0'], ['0']];
+        const shortAnswerQuestionUtil = TestBed.inject(ShortAnswerQuestionUtil);
+        spyOn(shortAnswerQuestionUtil, 'divideQuestionTextIntoTextParts').and.returnValue(textParts);
+
+        const node = {} as Node;
+
+        const returnValue = ({
+            contains(other: Node | null): boolean {
```
Comment (FrankeLukas) [inline src/test/javascript/spec/component/short-answer-question/short-answer-question-edit.component.spec.ts:178]: `'other' is defined but never used` there are a bunch more warnings we should at least ignore with that  `// tslint ignore` pattern unless we can't fix them

**Review by FrankeLukas — CHANGES_REQUESTED** [review]

**Comment by Strohgelaender** [comment]
I don't think these issues are already resolved, it happened again just yesterday (4.10.3). An Instructor overwrote the assessment which already had an accepted complaint and now the complaint is gone (hasComplaint: false). Can you please check this again?

File: src/test/javascript/spec/component/short-answer-question/short-answer-question-edit.component.spec.ts (hunk 1 above)
Comment (FuchsDominik) [reply src/test/javascript/spec/component/short-answer-question/short-answer-question-edit.component.spec.ts:179]: 👍  55aea5a

**Comment by FuchsDominik** [comment]
> I don't think these issues are already resolved, it happened again just yesterday (4.10.3) for a programming exercise. An Instructor overwrote the assessment which already had an accepted complaint and now the complaint is gone (hasComplaint: false). Can you please check this again?
> 
> Edit: It seems like this is fixed for Modeling Exercises (see [AssessmentService.java#L347](https://github.com/ls1intum/Artemis/blob/749fa50c416ba1e5a9b8cc739a4fe32dca8c27e8/src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java#L347)) but not for programming exercises (see [ProgrammingAssessmentService.java#L44](https://github.com/ls1intum/Artemis/blob/749fa50c416ba1e5a9b8cc739a4fe32dca8c27e8/src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingAssessmentService.java#L44)). These tests only check modeling and not programming exercises.

@Strohgelaender Thanks for telling! I overlooked the [note](https://github.com/ls1intum/Artemis/blob/749fa50c416ba1e5a9b8cc739a4fe32dca8c27e8/src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java#L347). I pushed a fix which should resolve this issue. As you already mentioned, the bug only existed for programming exercises, where it was just assumed that there is no complaint. I added a check whether the prior result had a complaint, which is then stored in the new result.

**Comment by Strohgelaender** [comment]
I'm not sure about this because I don't know the details about how the assessments are handled in Artemis, but doesn't this change the behavior on new assessments? hasComplaint is from Type Boolean and null-values are also possible here. So it will stay at null here hence it's always set to the old value, right? The AssesmentService contains other logic, here it will be set to false if the previous value is null/empty.

**Comment by FuchsDominik** [comment]
> I'm not sure about this because I don't know the details about how the assessments are handled in Artemis, but doesn't this change the behavior on new assessments? hasComplaint is from Type Boolean and null-values are also possible here. So it will stay at null here hence it's always set to the old value, right? The AssesmentService contains other logic, here it will be set to false if the previous value is null/empty.

@Strohgelaender It's true that the other AssessmentService contains other logic. IMO my implementation just copies the value stored in the latest Result, which is what we want at this point, I guess. If the hasComplaint flag is true, the new value will be true, if it is false the new value will be false und if it is null the new value will be null. It seems that for modeling-/... exercises hasComplaint cannot be null, whereas for programming exercises it can be null. If we would like to change this behaviour we would need such a check like in the AssessmentService and like you mentioned.

**Comment by Strohgelaender** [comment]
The difference is, that in the old behavior the value can never be null again after the manual assessment. Now it will stay null. But as I said I don't know any details about the implementation here, this may be exactly what is wanted, I just wanted to point out that this introduces some (maybe unwanted) behavior changes. 🤷‍♂️

**Review by airelawaleria — COMMENTED** [review]
So basically the complaint hasn't disappeared from the assessment dashboard for other users, but it got locked by the instructor and there was no possibility to unlock it, since I, as an instructor, have overwritten it from the Submissions page and it wasn't shown/added in the assessment dashboard for me as an instructor

**Comment by krusche** [comment]
I would propose that we get rid of the `null` case at all. The default value for the database column should be `false` and it can only be set to true, when the student creates a complaint.
(We do not have a delete complaint functionality as far as I know)

This would solve further issues with the 3rd case `null` which would not exist any more

**Comment by FuchsDominik** [comment]
Ok so the logic when saving a manual programming assessment is that it first checks whether or not the old hasComplaint flag is set. If it's not set (so null I assume), hasComplaint will be set to false. If it is set, it checks whether or not hasComplaint is true or false and sets the new hasComplaint value for the newly created result accordingly. Do you agree on this behaviour?

**Comment by tiffca** [comment]
For me the assessment gets locked by the instructor and I can't access the complaint anymore (as a different tutor) - also there is no way for me to unlock the assessment as the instructor. Expected: the complaint should be unlocked after the assessment is overridden by the instructor/I should be able to evaluate the complaint as a different tutor. 

image

**Comment by Strohgelaender** [comment]
> 
> 
> Ok so the logic when saving a manual programming assessment now is that it first checks whether or not the old hasComplaint flag is set. If it's not set (so null I assume), hasComplaint will be set to false. If it is set, it checks whether or not hasComplaint is true or false and sets the new hasComplaint value for the newly created result accordingly. Do you agree on this behaviour? @krusche @Strohgelaender

This sounds consistent with the old logic and also with the behavior of the other exercise types. 👍But as mentioned by @krusche it seems better to remove the null case completely, maybe on a follow-up-PR.

**Review by philgabr — COMMENTED** [review]
I can confirm the aforementioned complaint lock problems. 

Also the "progress statistic" showed 1 compliant, however below it says "there are no complaints at the moment" (ts3):

image

**Comment by FuchsDominik** [comment]
> I can confirm the aforementioned complaint lock problems.
> 
> Also the "progress statistic" showed 1 compliant, however below it says "there are no complaints at the moment" (ts3):
> 
> image

@philgabr Thanks for testing. I just check on ts3 where another branch is currently deployed and also found this scenario. 
Screenshot 2021-03-04 at 22 26 59
So this does not seem to be a problem connected to this particular PR, but i am still not quite sure why this behaviour occurs

**Comment by krusche** [comment]
Could it be the case that complaints about yourself are not shown in the list below? But are still counted for the overall progress

**Comment by FuchsDominik** [comment]
> Could it be the case that complaints about yourself are not shown in the list below? But are still counted for the overall progress

Oh yeah i overlooked that there is only one submission which @philgabr assessed himself. Like mentioned in the _?_ tooltip, you can only see complaints about assessments which are not your own, so this would probably be the case here. For my findings, it could be possible that this has something to do with creating submission with admin users which are not registered in the course, so more or less limited to the test server

**Comment by FuchsDominik** [comment]
> For me the assessment gets locked by the instructor and I can't access the complaint anymore (as a different tutor) - also there is no way for me to unlock the assessment as the instructor. Expected: the complaint should be unlocked after the assessment is overridden by the instructor/I should be able to evaluate the complaint as a different tutor.
> 
> Steps to reproduce:
> 
> 1. assess an assessment as a tutor
> 2. submit a complaint as a student
> 3. override the assessment as an instructor
> 4. try to evaluate the complaint as a different tutor
> 5. actual: can't evaluate. expected: should be able to evaluate
> 
> image

@tiffca Is it possible to somehow address/unlock the complaint by the instructor? Logically, if an instructor overrides an assessment, a tutor should not be able to change this assessment anymore, as the instructor might have had good reasons to do such an intervention and he would then be able to revert the instructors changes. This is also probably connected to @airelawaleria comment.

**Comment by FuchsDominik** [comment]
> So basically the complaint hasn't disappeared from the assessment dashboard for other users, but it got locked by the instructor and there was no possibility to unlock it, since I, as an instructor, have overwritten it from the Submissions page and it wasn't shown/added in the assessment dashboard for me as an instructor

Can you accept/reject the complaint by going into the assessment page via the submissions, like you also initially overwrote the assessment? It should not be shown at the bottom after these PR changes

**Comment by Strohgelaender** [comment]
The initial bug report addressed the following steps:

1. assess the assessment as a tutor
2. submit a complaint as a student
3. answer the complaint as another tutor
4. override the assessment as an instructor

actual: complaint details are gone
expected / now: complaint details are still there.
This is fixed by this PR.

Related to the steps mentioned above is #2027. If an Instructor was able to just answer a complaint regularly, there's no need to overwrite an assessment with an open complaint.

If the instructor nevertheless decides to do so, the best and easiest solution in my opinion would be to just mark the complaint as accepted. By this also no other tutor can overwrite the assessment of the instructor by answering the complaint.

This also leads to some follow-up-questions (which are probably out of scope for this PR):
1. how to deal with more feedback requests?
2. If an instructor overwrites an assessment without any complaint while the complaint deadline is not over yet, should the student still be able to write a complaint?

**Comment by FuchsDominik** [comment]
@Strohgelaender I agree. The problem mentioned by @tiffca and @airelawaleria originates in the fact that the assessor is not allowed to evaluate the complaint. Since the instructor locks the complaint by navigating into the assessment page and then overwrites the assessor/assessment, he/she is assessor of the submission **AND** the owner of the lock of this complaint. If we allow instructors to always be allowed to accept/reject complaints, the instructor could just accept/reject the complaint so the complaint is addressed. 
If I am not wrong here, a complaint lock should also be automatically unlocked in 24 hours, so at least this status would be solved 24 hours later :P
What i also wonder is, why would the instructor overwrite an assessment if he/she could just evaluate the complaint and therefore change the assessment.

File: src/test/java/de/tum/in/www1/artemis/service/ProgrammingExerciseScheduleServiceTest.java [hunk 2]
```diff
@@ -86,7 +86,7 @@ private void mockStudentRepoLocks() throws URISyntaxException {
     @Test
     void shouldExecuteScheduledBuildAndTestAfterDueDate() throws Exception {
         mockStudentRepoLocks();
-        long delayMS = 800;
+        long delayMS = 1000;
```
Comment (SimonEntholzer) [inline src/test/java/de/tum/in/www1/artemis/service/ProgrammingExerciseScheduleServiceTest.java:89]: Why was this change necessary? :D

**Review by SimonEntholzer — APPROVED** [review]
Testing on ts3 worked fine, the code also looks good :) one thing that is a bit odd:
 

After I override the assessment+complaint with a instructor the original assessor changes (which makes kind of sense). Still, this now "original assessor" is not really the original assessor.
Not sure though if there could be anything done about it.

**Review by tiffca — APPROVED** [review]
Tested successfully

File: src/test/java/de/tum/in/www1/artemis/service/ProgrammingExerciseScheduleServiceTest.java (hunk 2 above)
Comment (FuchsDominik) [reply src/test/java/de/tum/in/www1/artemis/service/ProgrammingExerciseScheduleServiceTest.java:89]: Like mentioned in the PR description, this was a flaky test which I tried to fix by giving it more time to exercute

**Review by iozbeyli — COMMENTED** [review]
Code looks good but there is a duplication warning in Codacy

**Review by FrankeLukas — APPROVED** [review]
